Aac file parsing

tenacex opened the pull request:
This PR includes support for AAC file parsing. It works with the existing mp4 transmuxer and can successfully output an audio segment. As per request, it currently does not support flv. The debug page has also been extended to support AAC file input, which it can detect automatically.

imbcmdth commented on debug/index.html:
```diff
@@ -138,6 +138,7 @@ <h1 class="title">Transmux Analyzer</h1>
     var inputs = document.getElementById('inputs'),
         original = document.getElementById('original'),
         working = document.getElementById('working'),
+        original_aac = document.getElementById('original-aac'),
```
Is this used anywhere?

tenacex replied:
Removed.

imbcmdth commented on debug/index.html:
```diff
@@ -326,7 +330,6 @@ <h1 class="title">Transmux Analyzer</h1>
 
       reader.addEventListener('loadend', function() {
         var bytes = new Uint8Array(reader.result);
-
```
You keep removing whitespace! The careful use of whitespace (in this case, after variable definitions) adds some readability to the code.

tenacex replied:
Re-added.

imbcmdth commented on lib/AACStream/stream-types.js:
```diff
@@ -0,0 +1,7 @@
+'use strict';
```
Why was this file copied to the AACStream directory?

tenacex replied:
Removed from the next commit.

imbcmdth commented on lib/AACStream/metadata-stream.js:
```diff
@@ -0,0 +1,229 @@
+/**
```
Why was this file copied to the AACStream directory?

tenacex replied:
This is a slightly altered metadata-stream since it looks for the timestamp ID3 tags and then sends them backward through the pipeline. I can however alter the file in the m2ts folder, but then I believe the file would be in an incorrect folder since it is being used by files outside the m2ts directory.

imbcmdth commented on lib/AACStream/aacs.js:
```diff
@@ -0,0 +1,123 @@
+/**
```
Following the naming conventions in the rest of the project, this file should be `aac.js` and be located in a directory called `aac`. You could also just make this file be the index.js in `aac`, I would be ok with that and it'll reduce the file count by one.

imbcmdth commented on lib/AACStream/aacs.js:
```diff
@@ -0,0 +1,123 @@
+// object types
+var AacStream;
+
+var
```
Because of the way `var` is being used here, you are defining AacStream twice. I suspect one is just leftover.

tenacex replied:
Changed.

imbcmdth commented on lib/AACStream/aacs.js:
```diff
@@ -0,0 +1,123 @@
+ AacStream = function() {
+  var
+    everything,
+    frameSize = 0,
```
Why is `frameSize` defined here if it is only used in `this.push`?

tenacex replied:
Moved inside of push.

imbcmdth commented on lib/AACStream/aacs.js:
```diff
@@ -0,0 +1,123 @@
+            canParseMore = false;
+          }
+        } else if ((everything[0] & 0xff == 0xff) &&
+            (((everything[1] >> 4) & 0xf) == 0xf)) {
```
While it doesn't matter here, you almost always want the zero-filling right shift `>>>` instead of the sign-extending right shift. This difference can bite you in subtle ways when dealing with large numbers and the zero-filling one is generally "safer". Also you can just mask the upper 4 bits and do: `(everything[1] & 0xF0) === 0xF0`.

tenacex replied:
Changed.

imbcmdth commented on lib/AACStream/aacs.js:
```diff
@@ -0,0 +1,123 @@
+          } else {
+            canParseMore = false;
+          }
+        } else if ((everything[0] & 0xff == 0xff) &&
```
You are using the unsafe type-coercing equality operator here. ALWAYS use `===` and forget you ever knew about `==` when doing JavaScript.

tenacex replied:
Changed.

imbcmdth commented on lib/AACStream/aacs.js:
```diff
@@ -0,0 +1,123 @@
+
+
+  this.push = function(bytes) {
+    var canParseMore = true;
```
Is there no way to write this loop without using a boolean variable?

tenacex replied:
Changed.

imbcmdth commented on lib/AACStream/aacs.js:
```diff
@@ -0,0 +1,123 @@
+  AacStream.prototype.init.call(this);
+
+  this.setTimestamp = function (frame) {
+    var d = frame.data;
```
You do this several times. Right now this project is using es5 conventions and our style guide proscribes that we only use **one** `var` statement at the top of the function always.

tenacex replied:
Changed.

imbcmdth commented on lib/AACStream/aacs.js:
```diff
@@ -0,0 +1,123 @@
+    var size = ((d[3] & 0x01)  << 30) | (d[4]  << 22) | (d[5] << 14) | (d[6] << 6) | d[7] >>> 2;
+    size *= 4;
+    size += d[7] & 0x03;
+    timeStamp = size;
```
I prefer it if functions didn't mutate outside state. This function should just return a value.

tenacex replied:
This function is the function that runs when we receive a 'timestamp' event. How can it be affective if it doesn't alter the state outside of it?

imbcmdth commented on lib/AACStream/aacs.js:
```diff
@@ -0,0 +1,123 @@
+    // If there are bytes remaining from the last segment, prepend them to the
+    // bytes that were pushed in
+    if (everything !== undefined && everything.length) {
+      var tempLength = everything.length;
```
This `var` should be at the top of the function.

tenacex replied:
changed.

imbcmdth commented on lib/AACStream/aacs.js:
```diff
@@ -0,0 +1,123 @@
+      if (everything.length >= 10) {
+        if ((everything[0] === 'I'.charCodeAt(0)) &&
+          (everything[1] === 'D'.charCodeAt(0)) &&
+          (everything[2] === '3'.charCodeAt(0))) {
```
I think this code would be much cleaner if ID3 tag parsing and ADTS frame parsing were separate functions called based on what the first few bytes are.

tenacex replied:
I pulled out the framesize calculations which took up a lot of space and made it hard to read. I left in the actual construction of the packet/chunk to keep with your mentality of not having functions modify outside of their scope. Let me know if you want it modularized further.

imbcmdth commented on lib/AACStream/aacs.js:
```diff
@@ -0,0 +1,123 @@
+    }
+  };
+  this.flush = function () {
+    this.trigger('done');
```
This trigger is performed implicitly if you don't define a `flush` function. Just delete this function.

imbcmdth commented on lib/AACStream/aacs.js:
```diff
@@ -0,0 +1,123 @@
+
+var AACStream = {
+  AacStream: AacStream,
+  MetadataStream: require('./metadata-stream'),
```
Why are requiring a different `metadata-stream.js` than the usual one?

tenacex replied:
This is a slightly altered metadata-stream since it looks for the timestamp ID3 tags and then sends them backward through the pipeline. I can however alter the file in the m2ts folder, but then I believe the file would be in an incorrect folder since it is being used by files outside the m2ts directory.

imbcmdth commented on lib/AACStream/aacs.js:
```diff
@@ -0,0 +1,123 @@
+    }
+
+    while(true) {
+      if (everything.length >= 10) {
```
Why not make the `while` condition equal to this `if`'s condition? Then you won't have the danger of infinite iteration looming over you! Also there should be a space between the keyword and the opening parens: `while (...`

tenacex replied:
Changed. Thanks for that tip.

imbcmdth commented on lib/mp4/transmuxer.js:
```diff
@@ -546,51 +547,103 @@ Transmuxer = function(options) {
-  // Hook up CEA-608/708 caption stream
-  h264Stream.pipe(captionStream)
-    .pipe(coalesceStream);
+    //audioSegmentStream = new AudioSegmentStream();
```
Leftover?

imbcmdth commented on lib/mp4/transmuxer.js:
```diff
@@ -546,51 +547,103 @@ Transmuxer = function(options) {
+  });
+  };
+
+    this.setupTSPipeline = function() {
```
Indenting here looks wonky.

tenacex replied:
Changed.

imbcmdth commented on lib/mp4/transmuxer.js:
```diff
@@ -546,51 +547,103 @@ Transmuxer = function(options) {
+
+
+  this.addAACListener = function() {
+      aacStream.on('data', function(data) {
```
Indenting looks wonky.

tenacex replied:
Changed.

imbcmdth commented on lib/mp4/transmuxer.js:
```diff
@@ -546,51 +547,103 @@ Transmuxer = function(options) {
+        adtsStream
+          .pipe(audioSegmentStream)
+          .pipe(coalesceStream);
+
```
There are extra empty newlines in this file... is this where all the ones that were missing in the other file snuck off to? :smile:

tenacex replied:
Changed... I think?

imbcmdth commented on lib/mp4/transmuxer.js:
```diff
@@ -638,13 +691,37 @@ Transmuxer = function(options) {
+
+  Transmuxer.prototype.init.call(this);
+  options = options || {};
+  aacfile = options.aacfile;
```
Does this need to be a variable? You can just do `option.aacfile === undefined` in the one place it is used. Also consider naming this something like `aacInput`.

tenacex replied:
It is used in 3 other places in this file. Don't you think it deserves variable status?

imbcmdth replied:
It would only be used once if you followed my two recommendations in the comments following this one. Also it should definitely have a name like `aacInput`

tenacex replied:
Just took care of all of this.

imbcmdth commented on lib/mp4/transmuxer.js:
```diff
@@ -663,13 +740,23 @@ Transmuxer = function(options) {
   // feed incoming data to the front of the parsing pipeline
   this.push = function(data) {
-    packetStream.push(data);
+    if (aacfile === undefined) {
```
You can get rid of this whole condition if you just saved whatever object was the head of the pipeline to a single variable. In other words, on push you should always push data into `pipelineRoot` or some such. Which object it actually is will be determined by which setup function was called.

imbcmdth commented on lib/mp4/transmuxer.js:
```diff
@@ -663,13 +740,23 @@ Transmuxer = function(options) {
   this.flush = function() {
     // Start at the top of the pipeline and flush all pending work
-    packetStream.flush();
+    if (aacfile === undefined) {
```
Same here.

imbcmdth commented on lib/AACStream/aacs.js:
```diff
@@ -0,0 +1,123 @@
+    if (footerpresent) {
+            return returnSize + 20;
+    }
+    return returnSize + 10;t
```
`return returnSize + 10;t` ??

tenacex replied:
Changed, snuck in there somehow.

imbcmdth commented on lib/AACStream/aacs.js:
```diff
@@ -0,0 +1,123 @@
+
+
+  this.push = function(bytes) {
+    var canParseMore = true,
```
Is `canParseMore` used anymore?

tenacex replied:
Removed.

imbcmdth commented on lib/AACStream/aacs.js:
```diff
@@ -0,0 +1,123 @@
+
+          //we have enough in the buffer to emit a full packet
+          if(everything.length >= frameSize) {
+            var chunk = {
```
Style guide specifies that you can only have one `var` at the top of functions.

imbcmdth commented on lib/AACStream/aacs.js:
```diff
@@ -0,0 +1,123 @@
+          frameSize = this.parseADTS(everything);
+
+          if(everything.length >= frameSize) {
+            var packet = {
```
Style guide specifies that you can only have one `var` at the top of functions.

imbcmdth commented on lib/AACStream/aacs.js:
```diff
@@ -0,0 +1,120 @@
+        frameSize = this.parseID3(everything);
+
+        //we have enough in the buffer to emit a full packet
+        if(everything.length >= frameSize) {
```
Missing a space after the `if`.

tenacex replied:
Changed.

imbcmdth commented on lib/AACStream/aacs.js:
```diff
@@ -0,0 +1,120 @@
+
+        frameSize = this.parseADTS(everything);
+
+        if(everything.length >= frameSize) {
```
Missing a space after the `if`.

tenacex replied:
Changed.

imbcmdth commented on lib/AACStream/aacs.js:
```diff
@@ -0,0 +1,120 @@
+        frameSize = this.parseADTS(everything);
+
+        if(everything.length >= frameSize) {
+          var packet = {
```
Style guide specifies that you can only have one var and it must appear at the top of functions.

tenacex replied:
Fixed.

imbcmdth commented on lib/AACStream/aacs.js:
```diff
@@ -0,0 +1,120 @@
+
+        //we have enough in the buffer to emit a full packet
+        if(everything.length >= frameSize) {
+          var chunk = {
```
Style guide specifies that you can only have one var and it must appear at the top of functions.

tenacex replied:
Fixed.

imbcmdth commented:
Missing tests!

tenacex commented:
Writing now.

dmlap commented on lib/AACStream/aacs.js:
```diff
@@ -0,0 +1,120 @@
+'use strict';
+var Stream = require('../utils/stream.js');
+
+var
```
Either everything below this needs to be indented differently or you should declare things here and then define them below.

tenacex replied:
Fixed. I decided to declare the AacStream above and then keep the indentation the same below.

dmlap commented on debug/index.html:
```diff
@@ -239,6 +239,7 @@ <h1 class="title">Transmux Analyzer</h1>
       codecsArray = codecs.value.split(',');
 
       mediaSource.addEventListener('sourceopen', function () {
+        mediaSource.duration = 0;
```
Why set the duration to zero here?

tenacex replied:
This allows the vide element to calculate the duration itself.

imbcmdth replied:
After the append, the duration will be updated by MSE if it is greater than the existing duration.

dmlap replied:
Setting it to zero is not truthful and could mislead downstream clients. This doesn't seem correct to me.

dmlap replied:
... keep forgetting this is the debug page. Less important but it still seems like it could mislead *us* at some point.

dmlap commented on lib/AACStream/aacs.js:
```diff
@@ -0,0 +1,120 @@
+
+  AacStream.prototype.init.call(this);
+
+  this.setTimestamp = function (frame) {
```
Calling this method setTimestamp() implies some behaviors that wouldn't be true in this case. It's a common convention to create "setters" that exist to allow downstream developers to mutate some internal properties of an object. In that case though, the setter would normally take as an argument the value that is being set. I think it's a bit confusing to call this setTimestamp() and have it take a frame, given that practice.

How about splitting the side-effect portion of this out from the calculate-timestamp-from-frame portion? Something like:
```js
this.calculateTimestamp_ = function(frame) {
  /* ... */ 
};

// later, when you need to set the timestamp:
timestamp = this.calculateTimestamp_(frame);
```

tenacex replied:
The thing is that it is taking the value that is being set. It receives an ID3 tag that I then use to set the `timestamp` variable. Then whenever we emit an ADTS frame we check the timestamp value and add in the necessary info. It is setup like this because we can have an arbitrary amount of ADTS frames following an ID3 tag. I think we should discuss alternatives in person.

tenacex replied:
Changed. See commit message.

dmlap commented on lib/AACStream/aacs.js:
```diff
@@ -0,0 +1,120 @@
+    timeStamp = size;
+  };
+
+  this.parseID3 = function(everything) {
```
Shadowing `everything` here may confuse some people.

tenacex replied:
Changed to `header`

dmlap commented on lib/AACStream/aacs.js:
```diff
@@ -0,0 +1,120 @@
+    var returnSize = (everything[6] << 21) | (everything[7] << 14) |
+              (everything[8] << 7) | (everything[9]),
+      flags = everything[5],
+      footerpresent = (flags & 16) >> 4;
```
variablesShouldBeCamelCased

tenacex replied:
iChangedThis.

imbcmdth commented:
There are still changes that need to be made here:
* Don't include `debug/mse-workbench.html` in this PR
* Rename `lib/AACStream` directory to just `lib/aac`
* Remove `lib/AACStream/metadata-stream.js`
* Remove `lib/AACStream/stream-types.js`
* Rename `lib/AACStream/aacs.js` to `lib/aac/index.js`
* Remove `test/files/trim.aac`

tenacex commented:
@imbcmdth & @dmlap I believe I have addressed most of your comment on this PR. Let me know if you have any more adjustments that you want made, and if not I can squash these commits and open up a new PR.

imbcmdth commented on lib/aac/index.js:
```diff
@@ -0,0 +1,121 @@
+    timeStamp = timestamp;
+  };
+
+  this.parseID3 = function(header) {
```
I think a name such as `parseID3TagSize` might be a more accurate name.

tenacex replied:
Changed.

imbcmdth commented on lib/aac/index.js:
```diff
@@ -0,0 +1,121 @@
+    return returnSize + 10;
+  };
+
+  this.parseADTS = function(header) {
```
I think a name such as `parseAdtsSize` might be a more accurate name.

tenacex replied:
Changed.

imbcmdth commented on lib/aac/index.js:
```diff
@@ -0,0 +1,121 @@
+          everything = everything.subarray(frameSize);
+          continue;
+        } else {
+          break;
```
You can flip the condition above and make this shorter:

````javascript
if (frameSize > everything.length) {
  break;
}
// ...everything else..
```

tenacex replied:
Changed.

imbcmdth commented on lib/aac/index.js:
```diff
@@ -0,0 +1,121 @@
+          everything = everything.subarray(frameSize);
+          continue;
+        } else {
+          break;
```
You can flip the condition above here too.

tenacex replied:
Changed.

imbcmdth commented on lib/aac/index.js:
```diff
@@ -0,0 +1,121 @@
+          break;
+        }
+      }
+      break;
```
This is strange. It says, "we aren't in an ID3 tag or a ADTS frame so break and wait for more data." We try to be more resilient to errors than this. In my opinion, we should walk forward a byte and see if find a valid id3 tag or adts frame there.

tenacex replied:
Changed. I removed the break and set `everything

imbcmdth commented on lib/aac/index.js:
```diff
@@ -0,0 +1,121 @@
+
+var AACStream = {
+  AacStream: AacStream,
+  MetadataStream: require('../m2ts/metadata-stream'),
```
There is no reason to require this and re-export it.

tenacex replied:
Changed.

imbcmdth commented on lib/aac/index.js:
```diff
@@ -0,0 +1,121 @@
+
+AacStream.prototype = new Stream();
+
+var AACStream = {
```
If you get rid of the MetadataStream export, you can get rid of this whole object and just export `AacStream` directly.

tenacex replied:
Changed.

imbcmdth commented on lib/aac/index.js:
```diff
@@ -0,0 +1,121 @@
+
+        //check framesize
+        frameSize = this.parseID3(everything);
+
```
Extra line.

imbcmdth commented on lib/aac/index.js:
```diff
@@ -0,0 +1,121 @@
+        //check framesize
+        frameSize = this.parseID3(everything);
+
+
```
Extra blank line.

imbcmdth commented on lib/m2ts/metadata-stream.js:
```diff
@@ -74,11 +74,14 @@ var
+
       tag.privateData = tag.data.subarray(i + 1);
       tag.data = tag.privateData;
+
```
These three changes are just added newlines. Remove them.

tenacex replied:
Changed.

imbcmdth commented on lib/m2ts/metadata-stream.js:
```diff
@@ -113,8 +116,8 @@ MetadataStream = function(options) {
 
   this.push = function(chunk) {
-    var tag, frameStart, frameSize, frame, i;
 
```
This is also simply an extra newline.

imbcmdth commented on lib/m2ts/metadata-stream.js:
```diff
@@ -194,17 +198,27 @@ MetadataStream = function(options) {
-
-      frame = {
-        id: String.fromCharCode(tag.data[frameStart],
+      var frameHeader = String.fromCharCode(tag.data[frameStart],
```
Once again: only define variables at the top of files.

tenacex replied:
Changed.

imbcmdth commented on lib/mp4/transmuxer.js:
```diff
@@ -546,104 +547,170 @@ Transmuxer = function(options) {
+    coalesceStream,
+    aacfile;
+
+  this.setupAACPipeline = function() {
```
Our caps standard for this project means that this should be called `setupAacPipeline`.

tenacex replied:
Changed.

imbcmdth commented on lib/mp4/transmuxer.js:
```diff
@@ -546,104 +547,170 @@ Transmuxer = function(options) {
-  options = options || {};
 
-  this.baseMediaDecodeTime = options.baseMediaDecodeTime || 0;
+  this.addAACListener = function() {
```
Our caps standard for this project means that this should be called `addAacListener `.

tenacex replied:
Changed.

imbcmdth commented on lib/mp4/transmuxer.js:
```diff
@@ -546,104 +547,170 @@ Transmuxer = function(options) {
-  // Hook up CEA-608/708 caption stream
-  h264Stream.pipe(captionStream)
-    .pipe(coalesceStream);
+    this.setupTSPipeline = function() {
```
Our caps standard for this project means that this should be called `setupTsPipeline`

tenacex replied:
Changed.

imbcmdth commented on lib/mp4/transmuxer.js:
```diff
@@ -546,104 +547,170 @@ Transmuxer = function(options) {
-        } else if (!audioTrack && data.tracks[i].type === 'audio') {
-          audioTrack = data.tracks[i];
-          audioTrack.timelineStartInfo.baseMediaDecodeTime = self.baseMediaDecodeTime;
+  this.addTSListener = function() {
```
Our caps standard for this project means that this should be called `addTsListener`

tenacex replied:
Changed.

imbcmdth commented on test/transmuxer.test.js:
```diff
@@ -2484,3 +2488,59 @@ QUnit.test('buffers video samples until flushed', function() {
   QUnit.equal(segments[0].tags.videoTags.length, 2, 'generated two video tags');
 });
+
+QUnit.test('AAcStream parses correct ID3 tag size', function() {
```
All of these tests should be in a `QUnit.module` named "AAC Stream". Then you can remove "AAcStream" from the test names.

BrandonOCasey commented on lib/aac/stream-types.js:
```diff
@@ -0,0 +1,7 @@
+'use strict';
+
+module.exports = {
```
Not sure if we need this file here, as it should be the same as m2ts stream-types

tenacex replied:
Removed.

BrandonOCasey commented on lib/m2ts/metadata-stream.js:
```diff
@@ -194,17 +194,27 @@ MetadataStream = function(options) {
                                 tag.data[frameStart + 1],
                                 tag.data[frameStart + 2],
-                                tag.data[frameStart + 3]),
+                                tag.data[frameStart + 3]);
```
tag.data should all be aligned

tenacex replied:
Changed.

imbcmdth commented on lib/mp4/transmuxer.js:
```diff
@@ -546,104 +547,170 @@ Transmuxer = function(options) {
-  // Hook up CEA-608/708 caption stream
-  h264Stream.pipe(captionStream)
-    .pipe(coalesceStream);
+    this.setupTsPipeline = function() {
```
This is too deeply indented.

imbcmdth commented on lib/mp4/transmuxer.js:
```diff
@@ -546,104 +547,170 @@ Transmuxer = function(options) {
+
+
+
+
```
Why so many empty lines?
